### PR TITLE
feat(FEC-10709, FEC-10712): player visibility - Auto-pause when player is out of view, Autoplay only when player is in view

### PR DESCRIPTION
**Pull request by RoyBregman**
### Description of the Changes

Migrated intersection observer logic from visibility plugin to KP

- Removed threshold from plugin config and docs
- PLAYER_VISIBILITY_CHANGED event is removed and not dispatched from plugin anymore
- Floating is now observing using the new viewabilityManager
- Plugin config - removed visibility level. plugin config is now called floating with the same floating config options

Solves FEC-10709, FEC-10712

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/visibility.js [hunk 1]
```diff
@@ -184,6 +167,12 @@ class Visibility extends BasePlugin {
     this.eventManager.listen(this.player, this.player.Event.LEAVE_PICTURE_IN_PICTURE, () => {
       this._isInPIP = false;
     });
+    this._viewablityManager = new Utils.ViewabilityManager();
+    this._viewablityManager.observe(this._appTargetContainer, this.player.config.playback.visibilityThreshold / 100, (visible: boolean) => {
```
Comment (Yuvalke) [inline src/visibility.js:171]: I think we should consider opening a new config cause it's not so relevant to playback

Comment (Yuvalke) [reply src/visibility.js:171]: You use here the same parameters to observe as in playkit why don't you just listen to playkit event?
if you want to be able to observe another element with a different threshold from this plugin I would use it like that

**Review by Yuvalke — CHANGES_REQUESTED** [review]

File: docs/configuration.md [hunk 2]
```diff
@@ -2,67 +2,31 @@
...
-  - [FLOATING_PLAYER_STATE_CHANGED][8]
-  - [FLOATING_PLAYER_DISMISSED][9]
-
-## VisibilityConfigObject
-
-Visibility plugin configuration parameters are provided whenever a player instance is created.
-
-Type: [Object][10]
-
-### Parameters
```
Comment (Yuvalke) [inline docs/configuration.md:21]: why did you remove these docs?

File: src/visibility.js [hunk 3]
```diff
@@ -184,6 +164,11 @@ class Visibility extends BasePlugin {
     this.eventManager.listen(this.player, this.player.Event.LEAVE_PICTURE_IN_PICTURE, () => {
       this._isInPIP = false;
     });
+    this.player.viewabilityManager.observe(this._appTargetContainer, this.player.config.visibility.playerThreshold / 100, (visible: boolean) => {
```
Comment (Yuvalke) [inline src/visibility.js:167]: don't you want it to be able to config separately? I mean if the plugin doesn't have a config threshold it will take the player threshold the same for targetContainer?

File: src/visibility.js [hunk 4]
```diff
@@ -203,8 +188,6 @@ class Visibility extends BasePlugin {
     this._appTargetContainer = null;
     this._floatingContainer = null;
     this._floatingPoster = null;
```
Comment (Yuvalke) [inline src/visibility.js:190]: I think you need to destroy the viewabilityManager

**Review by Yuvalke — CHANGES_REQUESTED** [review]

File: src/visibility.js (hunk 3 above)
Comment (OrenMe) [reply src/visibility.js:167]: In general the entire reason to expose the visibility manager as a service is for any plugin to use it with its own thresholds, but in this case the floating plugin should have the exact player thresholds so we can avoid needing to set it on the plugin level and use the player.
We can use this maybe as an escape hatch for any future issue and do what @Yuvalke mentioned to use player defaults and only use plugin config if it was set by app.

File: src/visibility.js (hunk 4 above)
Comment (OrenMe) [reply src/visibility.js:190]: the viewability manager is a player service, it is in charge of destroying it.

Comment (Yuvalke) [reply src/visibility.js:190]: remove the specific handler from it? if the plugin is not necessary anymore it'll keep the listener

Comment (RoyBregman) [reply src/visibility.js:190]: When Kaltura is destroyed we call intersection observer disconnect which remove all listeners.
Is there a scenario where we destroy only a plugin? If so then we need to to add an unobserve to the plugin destroy.

File: src/visibility.js (hunk 3 above)
Comment (RoyBregman) [reply src/visibility.js:167]: I don't understand - Do you want have the playerThreshold as default in the viewabilityManager and only if someone passes it in the observe to use it?

File: docs/configuration.md (hunk 2 above)
Comment (RoyBregman) [reply docs/configuration.md:21]: It was generated by script. In general I removed the visibility level of the plugin config. We have the floating directly

File: src/visibility.js (hunk 4 above)
Comment (OrenMe) [reply src/visibility.js:190]: @RoyBregman we should have an unobserve API as we discussed

File: src/visibility.js (hunk 3 above)
Comment (OrenMe) [reply src/visibility.js:167]: exactly - specifically for floating - it needs to have exactly the same threshold as the player IIRC

File: src/visibility.js (hunk 1 above)
Comment (RoyBregman) [reply src/visibility.js]: It is not the same. In Kaltura i observe config.ui.targetId (The player itself - even after floating)
Here we observe config.targetId (The placeholder)

File: src/visibility.js (hunk 3 above)
Comment (RoyBregman) [reply src/visibility.js]: done

File: src/visibility.js (hunk 4 above)
Comment (RoyBregman) [reply src/visibility.js:192]: We have an unObserve, i just didn't call it cause Kaltura player destroys the plugin when it is destroyed.
Anyway I added the unObserve call also here.

**Review by OrenMe — APPROVED** [review]

**Review by Yuvalke — APPROVED** [review]
